### PR TITLE
Add support to create a SparkSession context

### PR DESCRIPTION
**Pull request by rmettu-rms**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](doc/contribution-guidelines.md#commit-message-format) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** (link exiting issues here : https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
Removed support to create SparkContext, SqlContext, and HiveContext, which are replaced with SparkSession context.

**New behavior :**
Add support to create SparkSession context 

**BREAKING CHANGES**
Removes SparkContext, SqlContext, and HiveContext. Maybe existing applications upgrading to latest version of Spark JobServer should have to extend from SparkSessionJob rather than NewSparkJob

### CODE REVIEW

**Comment by rmettu-rms** [comment]
@velvia @zeitos @noorul please take a look 
Couple of things - 
1. I didn't have the right setup to run python tests, maybe someone can help give me some instructions or run these tests for me
2. Making SparkSession a default context, is that a configuration option or is hidden somewhere in the code.

**Comment by hntd187** [comment]
I believe the Python and some Java contexts still want a SparkContext, so I don't think you can remove that one at least, which might be why the python jobs are breaking.

**Comment by rmettu-rms** [comment]
@hntd187 @velvia @zeitos @noorul  Added support to create SparkSession context in Python Api. All the tests are passing but CI is showing 1 pending check with no errors
please take look, hoping we can merge this soon

**Comment by rmettu-rms** [comment]
Fixed the python code guidelines check failure - but now unrelated test failed - didn't find an option to rebuild - maybe someone who has access can rebuild

**Comment by noorul** [comment]
@rmettu-rms restarted

**Comment by rmettu-rms** [comment]
@noorul Thanks. Build is successful

**Comment by noorul** [comment]
@rmettu-rms Code coverage has come down. It will be good, if it can be improved.

**Comment by rmettu-rms** [comment]
@noorul I reviewed the code coverage details, found that context factory classes are not getting invoked in test cases but we have similar test context factory classes which have Hive metastore configuration needed in local test setup. 
If you find a particular area that needs some test cases please let me know, I will try to take a look

**Comment by rmettu-rms** [comment]
@hntd187 @noorul can I get a review please whenever you get a chance

File: job-server/src/main/scala/spark/jobserver/JobManagerActor.scala [hunk 1]
```diff
@@ -230,7 +230,7 @@ class JobManagerActor(contextConfig: Config, daoActor: ActorRef)
       None
     }
 
-    val daoAskTimeout = Timeout(3 seconds)
+    val daoAskTimeout = Timeout(600 seconds)
```
Comment (hntd187) [inline job-server/src/main/scala/spark/jobserver/JobManagerActor.scala]: I don't think we should change this. Things should timeout normally if something goes wrong

File: job-server-extras/src/test/scala/spark/jobserver/python/PythonJobManagerSpec.scala [hunk 2]
```diff
@@ -22,29 +22,30 @@ class PythonJobManagerSpec extends ExtrasJobSpecBase(PythonJobManagerSpec.getNew
...
       uploadTestEgg("python-demo")
 
       manager ! JobManagerActor.StartJob(
         "python-demo",
-        "example_jobs.word_count.WordCountSparkJob",
+        "example_jobs.word_count.WordCountSparkSessionJob",
         ConfigFactory.parseString("""input.strings = ["a", "b", "a"]"""),
         errorEvents ++ syncEvents)
-      expectMsgPF(3 seconds, "Expected a JobResult or JobErroredOut message!") {
+      expectMsgPF(15 seconds, "Expected a JobResult or JobErroredOut message!") {
```
Comment (hntd187) [inline job-server-extras/src/test/scala/spark/jobserver/python/PythonJobManagerSpec.scala]: Again are you changing these to deal with flaky tests?

File: job-server/src/main/scala/spark/jobserver/JobManagerActor.scala (hunk 1 above)
Comment (rmettu-rms) [inline job-server/src/main/scala/spark/jobserver/JobManagerActor.scala]: Reverted back the change

File: job-server-extras/src/test/scala/spark/jobserver/python/PythonJobManagerSpec.scala (hunk 2 above)
Comment (rmettu-rms) [inline job-server-extras/src/test/scala/spark/jobserver/python/PythonJobManagerSpec.scala]: I was getting a timeout error on this but testing it with 3 seconds now I can no longer reproduce. Reverted the change

**Comment by hntd187** [comment]
This looks good to me then, I'm gonna wait for @velvia to give the final go ahead on this

**Comment by rmettu-rms** [comment]
@velvia can you please take a look at this before it gets out of sync

File: README.md [hunk 3]
```diff
@@ -334,6 +334,38 @@ Let's try running our sample job with an invalid configuration:
...
+object WordCountExampleSparkSession extends SparkSessionJob {
+  type JobData = Seq[String]
+  type JobOutput = collection.Map[String, Long]
+
+  override def runJob(sparkSession: SparkSession, runtime: JobEnvironment, data: JobData): JobOutput =
+    sparkSession.sparkContext.parallelize(data).countByValue
+
+  override def validate(sparkSession: SparkSession, runtime: JobEnvironment, config: Config):
+  JobData Or Every[ValidationProblem] = {
+    Try(config.getString("input.string").split(" ").toSeq)
```
Comment (velvia) [inline README.md]: You can use scalactic's `.toOr` implicit, would help here a little bit

File: job-server-extras/src/main/scala/spark/jobserver/SparkSessionJob.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  def runJob(sparkSession: SparkSession, runtime: JobEnvironment, data: JobData): JobOutput
+
+  override def runJob(sc: C, runtime: JobEnvironment, data: JobData): JobOutput
+   = runJob(sc.spark, runtime, data)
+
+  // Application should use this method to override
+  def validate(sparkSession: SparkSession, runtime: JobEnvironment, config: Config):
+      JobData Or Every[ValidationProblem]
+
+  override def validate(sc: C, runtime: JobEnvironment, config: Config):
```
Comment (velvia) [inline job-server-extras/src/main/scala/spark/jobserver/SparkSessionJob.scala]: Hm, is the wrapper really necessary?   I'll have more comments when I peek at more of the code

File: job-server-extras/src/main/scala/spark/jobserver/context/SessionContextFactory.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package spark.jobserver.context
+
+import com.typesafe.config.Config
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.sql.SparkSession
+import spark.jobserver.{ContextLike, SparkSessionJob}
+import spark.jobserver.api.SparkJobBase
+import spark.jobserver.util.SparkJobUtils
+
+case class SparkSessionContextLikeWrapper(spark: SparkSession) extends ContextLike {
```
Comment (velvia) [inline job-server-extras/src/main/scala/spark/jobserver/context/SessionContextFactory.scala]: I think there is a way to do this without using this wrapper.   We do this in the other ContextFactories.  Something like:

    val session = new SparkSession with ContextLike

Comment (rammettu) [inline job-server-extras/src/main/scala/spark/jobserver/context/SessionContextFactory.scala]: @velvia I tried that approach as my first try but it did not work as only object SparkSession is exposed and could not extend it with ContextLike trait

Comment (velvia) [inline job-server-extras/src/main/scala/spark/jobserver/context/SessionContextFactory.scala]: Ok got it.

**Comment by velvia** [comment]
@rmettu-rms once you resolve conflicts then we can merge, thanks.

**Comment by rmettu-rms** [comment]
@velvia @noorul checked in changes (merged with latest master branch) that should resolve the conflicts. For some reason I didn't see it kicked off a build, can you please take a look.

**Comment by noorul** [comment]
Still there are conflicts. You need to rebase against https://github.com/spark-jobserver/spark-jobserver

**Comment by rmettu-rms** [comment]
@noorul Thanks
@velvia resolved the conflicts, this is now ready to be merged.

**Comment by velvia** [comment]
Thanks @rmettu-rms this is now merged

**Comment by muthurajr** [comment]
@rmettu-rms I do see WordCountExampleSparkSession is not included in the job-server-tests. Is it not added for any specific reason?

**Comment by noorul** [comment]
@rmettu-rms @velvia I think this broke backward compatibility. I think we should not have removed old classes.

**Pull request merged**